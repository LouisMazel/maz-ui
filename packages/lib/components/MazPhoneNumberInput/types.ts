import type { CountryCallingCode, CountryCode, NationalNumber, NumberType } from 'libphonenumber-js'

import { type defaultLocales } from './default-locales'

export type Result = {
  isValid: boolean
  isPossible?: boolean
  countryCode?: CountryCode
  countryCallingCode?: CountryCallingCode
  nationalNumber?: NationalNumber
  type?: NumberType
  formatInternational?: string
  formatNational?: string
  uri?: string
  e164?: string
  rfc3966?: string
}

export type Translations = typeof defaultLocales

export type Country = {
  iso2: CountryCode
  dialCode: CountryCallingCode
  name: string
}

export interface IpWhoResponse {
  ip: string
  success: true
  type?: string
  continent?: string
  continent_code?: string
  country?: string
  country_code?: string
  region?: string
  region_code?: string
  city?: string
  latitude?: number
  longitude?: number
  is_eu: true
  postal?: string
  calling_code?: string
  capital?: string
  borders?: string
  flag: {
    img?: string
    emoji?: string
    emoji_unicode?: string
  }
  connection: {
    asn?: number
    org?: string
    isp?: string
    domain?: string
  }
  timezone: {
    id?: string
    abbr?: string
    is_dst: false
    offset?: number
    utc?: string
    current_time?: string
  }
}
