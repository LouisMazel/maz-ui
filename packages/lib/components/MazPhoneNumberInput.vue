<template>
  <div
    :id="instanceId"
    class="m-phone-number-input"
    :class="{
      '--no-flags': noFlags,
    }"
  >
    <button
      v-if="countryCode && !noFlags && !noCountrySelector"
      class="m-phone-number-input__country-flag maz-text-xl"
      tabindex="-1"
      type="button"
      @click="focusCountrySelector"
    >
      {{ localeToUnicodeFlag(countryCode) }}
    </button>

    <MazSelect
      v-if="!noCountrySelector && countryOptions"
      ref="CountrySelector"
      class="m-phone-number-input__select"
      :model-value="countryCode"
      option-value-key="iso2"
      option-label-key="name"
      option-input-value-key="dialCode"
      :max-list-width="250"
      :disabled="disabled"
      :color="color"
      :size="size"
      :list-position="listPosition"
      :search="!noSearch"
      :search-placeholder="locales.countrySelector.searchPlaceholder"
      :options="countryOptions"
      :error="error || (!!inputValue && !countryCode)"
      :hint="!!inputValue && !countryCode ? locales.countrySelector.error : undefined"
      :label="locales.countrySelector.placeholder"
      @update:model-value="onCountryChanged($event)"
      @focus="inputFocused = false"
    >
      <template #default="{ option, isSelected }">
        <div
          class="m-phone-number-input__select__item maz-flex maz-items-center maz-truncate"
          :class="{
            'm-phone-number-input__select__item--selected': isSelected,
          }"
        >
          <span v-if="!noFlags && typeof option.iso2 === 'string'" class="maz-mr-2 maz-text-lg">
            {{ localeToUnicodeFlag(option.iso2) }}
          </span>
          <span
            v-if="showCodeOnList"
            class="maz-w-10 maz-flex-none"
            :class="{ 'maz-text-muted': !isSelected }"
          >
            {{ option.dialCode }}
          </span>
          <span class="maz-flex-1 maz-truncate" :class="{ 'maz-font-semibold': isSelected }">
            {{ option.name }}
          </span>
        </div>
      </template>
    </MazSelect>

    <MazInput
      :id="id"
      ref="PhoneNumberInput"
      :model-value="inputValue"
      :label="inputPlaceholder"
      :disabled="disabled"
      :color="color"
      :error="error || (!!inputValue && !results?.isValid)"
      v-bind="$attrs"
      :size="size"
      icon-name="phone"
      type="tel"
      clearable
      class="m-phone-number-input__input maz-flex-1"
      :class="{
        '--border-radius': !noCountrySelector,
        '--error': error || !results?.isValid,
        '--focused': inputFocused,
      }"
      @focus="inputFocused = true"
      @blur="inputFocused = false"
      @update:model-value="onInputValueChanged($event)"
    />
  </div>
</template>

<script lang="ts" setup>
  import type { CountryCode } from 'libphonenumber-js'
  import type { Result, Translations } from './MazPhoneNumberInput/types'
  import type { Color, Position, Size } from './types'

  export type { Color, Size, Position, CountryCode, Result, Translations }

  import { truthyFilter } from './../modules/helpers/truthy-filter'
  import { useInstanceUniqId } from '../modules/composables/use-instance-uniq-id'

  import { defaultLocales } from './MazPhoneNumberInput/default-locales'

  import {
    computed,
    onMounted,
    type PropType,
    ref,
    watch,
    getCurrentInstance,
    nextTick,
    reactive,
  } from 'vue'

  import { localeToUnicodeFlag } from './../modules/helpers/locale-to-unicode-flag'
  import { useLibphonenumber, isCountryAvailable } from './MazPhoneNumberInput/use-libphonenumber'

  const {
    fetchCountryCode,
    browserLocale,
    getResultsFromPhoneNumber,
    getAsYouTypeFormat,
    getCountriesList,
    getPhoneNumberExample,
    sanitizePhoneNumber,
    loadPhoneNumberExamplesFile,
  } = useLibphonenumber()

  import MazInput from './MazInput.vue'
  import MazSelect from './MazSelect.vue'

  const emits = defineEmits([
    /** emitted when country or phone number changed
     * @returns data - {Result}
     */
    'update',
    /** emitted when country or phone number changed
     * @returns data - {Result}
     */
    'data',
    /** emitted when selected country changed */
    'country-code',
    /** Two way binding (v-model:model-value) - emitted when country changed */
    'update:model-value',
    /** Two way binding (v-model:country-code) - emitted when country changed */
    'update:country-code',
  ])

  const props = defineProps({
    modelValue: {
      type: String,
      validator: (prop: string) => {
        return typeof prop === 'string' || prop === undefined
      },
      default: undefined,
    },
    /** @deprecated */
    defaultPhoneNumber: { type: String, default: undefined },
    countryCode: {
      type: String as PropType<CountryCode | string>,
      default: undefined,
      validator: (code: string) => isCountryAvailable(code),
    },
    /** @deprecated - use country-code or v-model:country-code */
    defaultCountryCode: {
      type: String as PropType<CountryCode | string>,
      default: undefined,
      validator: (code: string) => isCountryAvailable(code),
    },
    id: { type: String, default: undefined },
    placeholder: { type: String, default: undefined },
    preferredCountries: { type: Array as PropType<CountryCode[]>, default: undefined },
    ignoredCountries: { type: Array as PropType<CountryCode[]>, default: undefined },
    onlyCountries: { type: Array as PropType<CountryCode[]>, default: undefined },
    translations: { type: Object as PropType<Translations>, default: undefined },
    listPosition: {
      type: String as PropType<Position>,
      default: 'bottom left',
      validator: (value: Position) => {
        return ['top', 'top right', 'top left', 'bottom', 'bottom right', 'bottom left'].includes(
          value,
        )
      },
    },
    color: { type: String as PropType<Color>, default: 'primary' },
    size: {
      type: String as PropType<Size>,
      default: 'md',
      validator: (value: string) => {
        return ['mini', 'xs', 'sm', 'md', 'lg', 'xl'].includes(value)
      },
    },
    noFlags: { type: Boolean, default: false },
    disabled: { type: Boolean, default: false },
    noExample: { type: Boolean, default: false },
    noSearch: { type: Boolean, default: false },
    noUseBrowserLocale: { type: Boolean, default: false },
    fetchCountry: { type: Boolean, default: false },
    noCountrySelector: { type: Boolean, default: false },
    showCodeOnList: { type: Boolean, default: false },
    error: { type: Boolean, default: false },
    customCountriesList: {
      type: Object as PropType<Record<CountryCode, string>>,
      default: undefined,
    },
    /**
     * Disabled auto-format as you type
     */
    noFormattingAsYouType: { type: Boolean, default: false },
    /**
     * locale of country list - Ex: "fr-FR"
     * @default browser locale
     */
    countryLocale: { type: String, default: undefined },
  })

  const instance = getCurrentInstance()
  const instanceId = useInstanceUniqId({
    componentName: 'MazPhoneNumberInput',
    instance,
    providedId: props.id,
  })

  const PhoneNumberInput = ref<typeof MazInput>()
  const CountrySelector = ref<typeof MazSelect>()
  const selectionRange = reactive<{
    start?: number | null
    end?: number | null
    cursorAtEnd?: boolean
  }>({
    start: 0,
    end: 0,
    cursorAtEnd: true,
  })

  const examplesFileLoaded = ref(false)
  const inputFocused = ref(false)

  const countries = computed(() => getCountriesList(props.countryLocale, props.customCountriesList))

  const countriesList = computed(() => {
    return countries.value?.filter((item) => !props.ignoredCountries?.includes(item.iso2))
  })

  const countriesFiltered = computed(() => {
    const countries = props.onlyCountries || props.preferredCountries
    return countries?.map(
      (country) => countriesList.value?.find((item) => item.iso2.includes(country)),
    )
  })

  const otherCountries = computed(() => {
    return countriesList.value?.filter((item) => !props.preferredCountries?.includes(item.iso2))
  })

  const countriesSorted = computed(() => {
    return props.preferredCountries
      ? [...(countriesFiltered.value ?? []), ...(otherCountries.value ?? [])]
      : props.onlyCountries
      ? countriesFiltered.value
      : countriesList.value
  })

  const countryOptions = computed(() => {
    return countriesSorted.value
      ?.map((country) => {
        return country
          ? {
              ...country,
              dialCode: `+${country.dialCode}`,
            }
          : undefined
      })
      .filter(truthyFilter)
  })

  const locales = computed(() => ({
    ...defaultLocales,
    ...props.translations,
  }))

  const inputPlaceholder = computed(() => {
    if (props.placeholder) {
      return props.placeholder
    }

    const defaultPlaceholder = locales.value.phoneInput.placeholder

    if (props.noExample || !examplesFileLoaded.value) {
      return defaultPlaceholder
    } else {
      const example = getPhoneNumberExample(countryCode.value)
      return results.value.isValid || !example
        ? defaultPlaceholder
        : `${locales.value.phoneInput.example} ${example}`
    }
  })

  const model = computed({
    get: () => props.modelValue || props.defaultPhoneNumber,
    set: (value) => {
      emits('update:model-value', value)
    },
  })

  const internalCountryCode = ref<CountryCode>()

  const countryCode = computed({
    get: () =>
      (props.countryCode || props.defaultCountryCode || internalCountryCode.value) as
        | CountryCode
        | undefined,
    set: (value) => {
      emits('country-code', value)
      emits('update:country-code', value)
      internalCountryCode.value = value
    },
  })

  const internalValue = ref<string | undefined>(model.value)

  const results = ref<Result>(
    getResultsFromPhoneNumber({
      phoneNumber: model.value,
      countryCode: countryCode.value,
    }),
  )

  const asYouTypeFormatted = ref<string>()

  const inputValue = computed<string>(() => {
    if (props.noFormattingAsYouType) {
      return internalValue.value ?? ''
    }

    return asYouTypeFormatted.value ?? internalValue.value ?? ''
  })

  async function setCountryFromIpWho() {
    const fetchedLocale = await fetchCountryCode()
    if (fetchedLocale) setCountryCode(fetchedLocale)
  }

  async function loadExamples() {
    try {
      if (!props.noExample && !examplesFileLoaded.value) {
        await loadPhoneNumberExamplesFile()
        examplesFileLoaded.value = true
      }
    } catch {
      console.error('[maz-ui](MazPhoneNumberInput) while loading phone number examples file')
    }
  }

  onMounted(async () => {
    await parseModel(model.value)

    await nextTick()

    if (props.fetchCountry && !countryCode.value) {
      await setCountryFromIpWho()
    }

    if (!props.defaultCountryCode && !props.noUseBrowserLocale && !countryCode.value) {
      setCountryCodeFromBrowserLocale()
    }

    await loadExamples()
  })

  watch(
    () => results.value,
    (newResults) => {
      emits('update', newResults)
      emits('data', newResults)
    },
    { immediate: true, deep: true },
  )

  watch(
    () => countryCode.value,
    (newCountryCode, oldCountryCode) => {
      if (newCountryCode && newCountryCode !== oldCountryCode) {
        setCountryCode(newCountryCode)
        onInputValueChanged(inputValue.value)
      }
    },
  )

  watch(
    () => model.value,
    async (newModel, oldModel) => {
      if (newModel !== oldModel) {
        parseModel(newModel)
      }
    },
  )

  watch(
    () => inputValue.value,
    async (newModel, oldModel) => {
      if (!props.noFormattingAsYouType && newModel && newModel !== oldModel) {
        const input = getPhoneNumberInput()
        if (
          input &&
          !results.value.isValid &&
          typeof selectionRange.start === 'number' &&
          !selectionRange.cursorAtEnd
        ) {
          const start = selectionRange.start
          const end = selectionRange.end
          setTimeout(() => {
            input.setSelectionRange(start, end ?? start)
          }, 0)
        }
      }
    },
  )

  function setCountryCodeFromBrowserLocale() {
    const { locale } = browserLocale()
    if (locale) {
      setCountryCode(locale)
    }
  }

  function updateResults(phoneNumber?: string, countryCode?: CountryCode) {
    results.value = getResultsFromPhoneNumber({
      phoneNumber,
      countryCode,
    })
  }

  async function onInputValueChanged(phoneNumber: string) {
    const input = getPhoneNumberInput()
    selectionRange.start = input?.selectionStart
    selectionRange.end = input?.selectionEnd
    selectionRange.cursorAtEnd = selectionRange.start
      ? selectionRange.start >= phoneNumber.length
      : true

    const sanitizedPhoneNumber = sanitizePhoneNumber(phoneNumber)

    internalValue.value = sanitizedPhoneNumber

    const newResults = getResultsFromPhoneNumber({
      phoneNumber: sanitizedPhoneNumber,
      countryCode: countryCode.value,
    })

    model.value = newResults.e164
  }

  function onCountryChanged(codeCountry: string | CountryCode) {
    updateResults(inputValue.value, codeCountry as CountryCode)

    setCountryCode(codeCountry, true)
  }

  function setCountryCode(selectedCountryCode: string | CountryCode, autoFocusInput = false) {
    try {
      const countryAvailable = isCountryAvailable(selectedCountryCode)

      if (countryAvailable) {
        countryCode.value = selectedCountryCode as CountryCode

        if (autoFocusInput) {
          focusPhoneNumberInput()
        }
      }
    } catch (error) {
      console.error(`[maz-ui](MazPhoneNumberInput) ${error}`)
    }
  }

  async function parseModel(newModel?: string) {
    updateResults(newModel, countryCode.value)

    await nextTick()

    if (props.noFormattingAsYouType) {
      internalValue.value = inputValue.value
    } else {
      asYouTypeFormatted.value =
        selectionRange.cursorAtEnd || results.value.isValid
          ? (internalValue.value = getAsYouTypeFormat(
              results.value.countryCode ?? countryCode.value,
              results.value.formatNational ?? internalValue.value,
            ))
          : internalValue.value
    }

    await nextTick()

    autoUpdateCountryCodeWithResults(results.value)
  }

  function autoUpdateCountryCodeWithResults(newResults: Result) {
    if (newResults.countryCode && countryCode.value !== newResults.countryCode) {
      setCountryCode(newResults.countryCode)
    }
  }

  function getPhoneNumberInput() {
    return PhoneNumberInput.value?.$el.querySelector('input') as HTMLInputElement | undefined
  }

  function focusCountrySelector() {
    CountrySelector.value?.$el.querySelector('input')?.focus()
  }

  async function focusPhoneNumberInput() {
    await nextTick()
    getPhoneNumberInput()?.focus()
  }
</script>

<style lang="postcss" scoped>
  .m-phone-number-input {
    @apply maz-relative maz-flex;

    &__country-flag {
      position: absolute;
      bottom: 2px;
      left: 15px;
      z-index: 4;
      outline: none;
      border: none;
      padding: 0;
      margin: 0;
      cursor: pointer;
    }

    &__input.--border-radius:deep(.m-input-wrapper) {
      @apply maz-rounded-l-none;

      margin-left: -2px;
    }

    &__select {
      @apply maz-w-36;

      &__item {
        @apply maz-w-full maz-text-sm;
      }

      &:deep(.m-select-input .m-input-wrapper) {
        @apply maz-rounded-r-none !important;
      }
    }

    &:not(.--no-flags) {
      .m-phone-number-input__select:deep(.m-input-wrapper) .m-select-input {
        @apply maz-pl-11 !important;
      }
    }

    &__input {
      &.--error,
      &.--focused {
        @apply maz-z-1;
      }
    }
  }
</style>
