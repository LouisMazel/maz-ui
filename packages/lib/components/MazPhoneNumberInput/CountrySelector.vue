<template>
  <div class="m-country-selector" :class="[props.class, { '--no-flags': noFlags }]" :style="style">
    <button
      v-if="modelValue && !noFlags"
      :id="`country-selector-flag-button-${id}`"
      class="m-country-selector__country-flag maz-text-xl"
      tabindex="-1"
      type="button"
      :class="{
        '--should-have-bottom-flag': locales.countrySelector.placeholder.length > 0,
      }"
      @click="focusCountrySelector"
    >
      <!--
        @slot Country selector flag
          @binding {String} country-code - current selected country code - Ex: `"FR"`
      -->
      <slot name="selector-flag" :country-code="modelValue">
        {{ countryCodeToUnicodeFlag(modelValue) }}
      </slot>
    </button>

    <MazSelect
      :id="`country-selector-${id}`"
      ref="CountrySelectorRef"
      :model-value="modelValue"
      v-bind="$attrs"
      class="m-country-selector__select"
      option-value-key="iso2"
      option-label-key="name"
      :option-input-value-key="countrySelectorDisplayName ? 'name' : 'dialCode'"
      :max-list-width="250"
      :disabled
      :color
      :size
      :error
      :list-position
      :item-height="38"
      :success
      :search="!noSearch"
      :search-placeholder="locales.countrySelector.searchPlaceholder"
      :options="countryOptions"
      :hint="!!phoneNumber && !modelValue ? locales.countrySelector.error : undefined"
      :label="locales.countrySelector.placeholder"
      :style="{
        width,
      }"
      :exclude-selectors="[`#country-selector-flag-button-${id}`, ...(excludeSelectors ?? [])]"
      @update:model-value="$emit('update:model-value', $event as CountryCode)"
    >
      <template #no-results>
        <slot name="no-results"> </slot>
      </template>
      <template #default="{ option, isSelected }">
        <div
          class="m-country-selector__select__item maz-flex maz-items-center maz-gap-1 maz-truncate"
          :class="{
            'm-country-selector__select__item--selected': isSelected,
          }"
        >
          <span v-if="!noFlags && typeof option.iso2 === 'string'" class="maz-text-lg">
            <!--
              @slot Country list flag
                @binding {String} country-code - country code of option - Ex: `"FR"`
                @binding {{ iso2: string; dialCode: string; name: string; }} option - country data
                @binding {Boolean} is-selected - `true` if option is selected
            -->
            <slot
              name="country-list-flag"
              :country-code="option.iso2"
              :option="option"
              :is-selected="isSelected"
            >
              {{ countryCodeToUnicodeFlag(option.iso2) }}
            </slot>
          </span>
          <span
            v-if="showCodeOnList"
            class="maz-w-9 maz-flex-none"
            :class="{ 'maz-text-muted': !isSelected }"
          >
            {{ option.dialCode }}
          </span>
          <span class="maz-flex-1 maz-truncate" :class="{ 'maz-font-semibold': isSelected }">
            {{ option.name }}
          </span>
        </div>
      </template>
    </MazSelect>
  </div>
</template>

<script lang="ts" setup>
  import { ref, computed, type HTMLAttributes, type ComponentPublicInstance } from 'vue'
  import type {
    Color,
    Size,
    Position,
    CountryCode,
    Translations,
    InjectedData,
  } from '../MazPhoneNumberInput.vue'
  import MazSelect from '../MazSelect.vue'
  import { countryCodeToUnicodeFlag } from '../../modules/helpers/country-code-to-unicode-flag'
  import { truthyFilter } from '../../modules/helpers/truthy-filter'
  import { useMazPhoneNumberInput } from './use-maz-phone-number-input'
  import { injectStrict } from '../../modules/helpers/inject-strict'

  const props = withDefaults(
    defineProps<{
      /** Style attribut of the component root element */
      style?: HTMLAttributes['style']
      /** Class attribut of the component root element */
      class?: HTMLAttributes['class']
      modelValue?: CountryCode
      id: string
      color: Color
      size: Size
      preferredCountries?: CountryCode[]
      ignoredCountries?: CountryCode[]
      onlyCountries?: CountryCode[]
      customCountriesList?: Record<CountryCode, string>
      locales: Translations
      listPosition?: Position
      noFlags?: boolean
      noSearch?: boolean
      disabled?: boolean
      showCodeOnList?: boolean
      countryLocale?: string
      success?: boolean
      error?: boolean
      countrySelectorDisplayName?: boolean
      width: string
      excludeSelectors?: string[]
    }>(),
    {
      class: undefined,
      style: undefined,
      modelValue: undefined,
      listPosition: 'bottom left',
      preferredCountries: undefined,
      ignoredCountries: undefined,
      onlyCountries: undefined,
      customCountriesList: undefined,
      countryLocale: undefined,
      width: '9rem',
      excludeSelectors: undefined,
    },
  )

  defineEmits<{
    (event: 'update:model-value', countryCode?: CountryCode): void
  }>()

  const { phoneNumber } = injectStrict<InjectedData>('data')

  const CountrySelectorRef = ref<ComponentPublicInstance<typeof MazSelect>>()

  const { getCountriesList } = useMazPhoneNumberInput()

  const countries = computed(() => getCountriesList(props.countryLocale, props.customCountriesList))

  const countriesList = computed(() =>
    countries.value?.filter((item) => !props.ignoredCountries?.includes(item.iso2)),
  )

  const countriesFiltered = computed(() => {
    const countries = props.onlyCountries || props.preferredCountries
    return countries?.map((country) =>
      countriesList.value?.find((item) => item.iso2.includes(country)),
    )
  })

  const otherCountries = computed(() =>
    countriesList.value?.filter((item) => !props.preferredCountries?.includes(item.iso2)),
  )

  const countriesSorted = computed(() =>
    props.preferredCountries
      ? [...(countriesFiltered.value ?? []), ...(otherCountries.value ?? [])]
      : props.onlyCountries
        ? countriesFiltered.value
        : countriesList.value,
  )

  const countryOptions = computed(() =>
    countriesSorted.value
      ?.map((country) =>
        country
          ? {
              ...country,
              dialCode: `+${country.dialCode}`,
            }
          : undefined,
      )
      .filter(truthyFilter),
  )

  async function focusCountrySelector() {
    CountrySelectorRef.value?.$el.querySelector('input')?.focus()
  }
</script>

<style lang="postcss" scoped>
  .m-country-selector {
    @apply maz-relative;

    &__country-flag {
      position: absolute;
      left: 13px;
      z-index: 4;
      outline: none;
      border: none;
      padding: 0;
      margin: 0;
      cursor: pointer;

      &.--should-have-bottom-flag {
        bottom: 2px;
      }
    }

    &__select {
      &:deep(.m-input-label) {
        @apply !maz-p-0;
      }

      &__item {
        @apply maz-w-full maz-text-sm;
      }
    }

    &:not(.--no-flags) {
      .m-country-selector__select:deep(.m-select-input input) {
        @apply !maz-pl-10;
      }
    }
  }

  /* RESPONSIVE */
  .m-phone-number-input {
    &.--responsive .m-country-selector {
      @apply maz-min-w-full mob-m:maz-min-w-[inherit];

      &__select {
        @apply maz-min-w-full mob-m:maz-min-w-[inherit];

        :deep(.m-select-input .m-input-wrapper) {
          @apply maz-rounded-b-none mob-m:maz-rounded-b mob-m:maz-rounded-r-none;
        }
      }
    }

    &.--row .m-country-selector {
      &__select {
        :deep(.m-select-input .m-input-wrapper) {
          @apply maz-rounded-r-none;
        }
      }
    }

    &.--col .m-country-selector {
      @apply maz-min-w-full;

      &__select {
        @apply maz-min-w-full;

        :deep(.m-select-input .m-input-wrapper) {
          @apply maz-rounded-b-none maz-rounded-tr;
        }
      }
    }
  }
</style>
