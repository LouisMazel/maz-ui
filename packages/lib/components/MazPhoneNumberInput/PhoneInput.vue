<template>
  <MazInput
    :id="`phone-number-${id}`"
    ref="PhoneInputRef"
    :model-value="modelValue"
    v-bind="$attrs"
    :label="inputLabel"
    :disabled
    :color
    :error
    :size
    :success
    type="tel"
    inputmode="tel"
    class="m-phone-input"
    :class="{
      '--border-radius': hasRadius,
      '--error': error || !results.isValid,
      '--focused': inputFocused,
    }"
    @focus="inputFocused = true"
    @blur="inputFocused = false"
    @update:model-value="($event) => onValueChange($event as string)"
  />
</template>

<script lang="ts" setup>
  import { ref, computed, onMounted, type ComponentPublicInstance, nextTick } from 'vue'
  import MazInput from '../MazInput.vue'

  import type { Color, Size, Translations } from '../MazPhoneNumberInput.vue'
  import { useLibphonenumber } from './use-libphonenumber'
  import { useMazPhoneNumberInput } from './use-maz-phone-number-input'

  const props = withDefaults(
    defineProps<{
      id: string
      color: Color
      size: Size
      locales: Translations
      label?: string
      noExample?: boolean
      disabled?: boolean
      hasRadius?: boolean
      success?: boolean
      error?: boolean
      autoFormat?: boolean
      noFormattingAsYouType?: boolean
    }>(),
    {
      class: undefined,
      style: undefined,
      label: undefined,
    },
  )

  const emits = defineEmits<{
    (event: 'update:model-value', phoneNumber?: string): void
  }>()

  const modelValue = defineModel<string>()

  const { loadPhoneNumberExamplesFile, getPhoneNumberExample, examples } = useLibphonenumber()
  const { selectedCountry, results, saveCursorPosition } = useMazPhoneNumberInput()

  const inputFocused = ref(false)

  const inputLabel = computed(() => {
    if (props.label) {
      return props.label
    }

    const defaultPlaceholder = props.locales.phoneInput.placeholder

    if (props.noExample || !examples.value) {
      return defaultPlaceholder
    } else {
      const example = getPhoneNumberExample(selectedCountry.value)
      return results.value?.isValid || !example
        ? defaultPlaceholder
        : `${props.locales.phoneInput.example} ${example}`
    }
  })

  const PhoneInputRef = ref<ComponentPublicInstance>()

  async function loadExamples() {
    try {
      return loadPhoneNumberExamplesFile()
    } catch (error) {
      console.error('[maz-ui](MazPhoneNumberInput) while loading phone number examples file', error)
    }
  }

  async function onValueChange(value?: string) {
    if (PhoneInputRef.value && value) {
      saveCursorPosition(PhoneInputRef.value, value)
    }
    await nextTick()
    emits('update:model-value', value)
  }

  onMounted(() => {
    if (!props.noExample) {
      loadExamples()
    }
  })
</script>

<style lang="postcss" scoped>
  .m-phone-input {
    @apply maz-min-w-52 maz-flex-1;

    &.--border-radius:deep(.m-input-wrapper) {
      @apply maz-rounded-l-none;

      margin-left: -2px;
    }

    &.--error,
    &.--focused {
      @apply maz-z-1;
    }
  }
</style>
