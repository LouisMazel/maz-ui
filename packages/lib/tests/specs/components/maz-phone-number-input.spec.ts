import { mount, type VueWrapper } from '@vue/test-utils'
import MazPhoneNumberInput from '@components/MazPhoneNumberInput.vue'

describe('components/MazPhoneNumberInput.vue', () => {
  expect(MazPhoneNumberInput).toBeTruthy()

  let wrapper: VueWrapper<InstanceType<typeof MazPhoneNumberInput>>

  beforeEach(() => {
    wrapper = mount(MazPhoneNumberInput, {
      props: {
        modelValue: '+33658584729',
      },
    })
  })

  test('Should have an uniq id', async () => {
    expect(wrapper.find('#MazPhoneNumberInput-1').exists()).toBe(true)
  })

  test('Should have the provided ifd', async () => {
    wrapper = mount(MazPhoneNumberInput, {
      props: {
        id: 'test',
      },
    })

    expect(wrapper.find('#test').exists()).toBe(true)
  })

  test('Should have the good values with FR number', async () => {
    const inputElement = wrapper.find('#MazPhoneNumberInput-28').element as HTMLInputElement

    expect(inputElement.value).toBe('06 58 58 47 29')
    expect(wrapper.vm.modelValue).toBe('+33658584729')
    expect(wrapper.emitted('country-code')?.[0][0]).toBe('FR')
  })

  test('Should have the good values with BE number', async () => {
    await wrapper.setProps({
      modelValue: '+326453',
    })

    const inputElement = wrapper.find('#MazPhoneNumberInput-37').element as HTMLInputElement

    expect(inputElement.value).toBe('+32 64 53')

    expect(wrapper.vm.modelValue).toBe('+326453')
    expect(wrapper.emitted('country-code')?.[1][0]).toBe('BE')
  })
})
