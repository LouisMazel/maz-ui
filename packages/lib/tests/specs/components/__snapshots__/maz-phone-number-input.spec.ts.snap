// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`components/MazPhoneNumberInput.vue > Should match with the snapshot 1`] = `
"<div data-v-4fe76c65=\\"\\" id=\\"MazPhoneNumberInput-1\\" class=\\"m-phone-number-input\\">
  <!--v-if-->
  <maz-select-stub data-v-4fe76c65=\\"\\" style=\\"width: 9rem;\\" class=\\"m-phone-number-input__select\\" options=\\"[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]\\" optionvaluekey=\\"iso2\\" optionlabelkey=\\"name\\" optioninputvaluekey=\\"dialCode\\" listposition=\\"bottom left\\" required=\\"false\\" disabled=\\"false\\" open=\\"false\\" color=\\"primary\\" itemheight=\\"40\\" maxlistheight=\\"240\\" maxlistwidth=\\"250\\" size=\\"md\\" search=\\"true\\" searchplaceholder=\\"Search the country\\" multiple=\\"false\\" error=\\"true\\" hint=\\"Choose country\\" label=\\"Country code\\" success=\\"true\\"></maz-select-stub>
  <maz-input-stub data-v-4fe76c65=\\"\\" class=\\"m-phone-number-input__input maz-flex-1 --border-radius\\" modelvalue=\\"06 58 58 47 29\\" color=\\"primary\\" label=\\"Phone number\\" name=\\"input\\" type=\\"tel\\" required=\\"false\\" disabled=\\"false\\" error=\\"false\\" success=\\"true\\" warning=\\"false\\" noborder=\\"false\\" noradius=\\"false\\" size=\\"md\\" debounce=\\"false\\" debouncedelay=\\"500\\" validbutton=\\"false\\" validbuttonloading=\\"false\\" autofocus=\\"false\\" borderactive=\\"false\\" icon-name=\\"phone\\" clearable=\\"\\"></maz-input-stub>
</div>"
`;
